### DISCUSSION TITLE
@AutoService and ServiceLoader

### DISCUSSION THREAD

**Discussion by dhpagani** [open-telemetry/opentelemetry-java-instrumentation#9779] [Q&A] [answered]
Hi guys!

I'm creating a extension to define ratio based on attributes. The strucutre of the Samplers, ConfigurableSamplerProvider, and others are working fine.

Now, I'm trying to create some classes to group some filter rules and everything is working fine, except for the part to get the implementations of MyInterface. I'm using Goolge AutoService to help to register the classes and trying to get those implementations using ServiceLoader.

Below is the structured code :

```
public interface MyInterface {
	String getName();
}

@AutoService(MyInterface.class)
MyCustomClass1 implements MyInterface{}

@AutoService(MyInterface.class)
MyCustomClass2 implements MyInterface{}

@AutoService(ConfigurableSamplerProvider.class)
public class MySampler implements ConfigurableSamplerProvider {
	private static final Logger logger = Logger.getLogger(MySampler.class.getName());
	@Override
	public Sampler createSampler(ConfigProperties config) {
		ServiceLoader<MyInterface> loader = ServiceLoader.load(MyInterface.class);
                System.out.println("test"); /*output to console*/
		loader.forEach(a -> {
			System.out.println("MyInterface  " + a.getName()); /*no output*/
		});

		return Sampler.alwaysOn();
	}
	@Override
	public String getName() {
		return "MySampler";
	}
}
```
In the console, I have "test" string but none "My Interface XXX" , and the method getName() is called, but Otel works fine getting MySampler.

I also tried to use SafeServiceLoader, with  Thread.currentThread().getContextClassLoader() as a classesLoader parameter, but 
I think is someting related about the [way of working with classloaders](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/javaagent-structure.md), but I have no glue how to fix it.

In summary, i'm not able to use `@AutoService` with MyInterface, ServiceLoader didn't return my implementations. If I open the JAR in build/libs/*.jar, I have the folder /META-INF/services/a.b.c.myinterfaces with the correct name of MyCustomClass1 and MyCustomClass2.

Thank you !

**Comment by laurit** [comment, marked as answer]
I think your hunch is correct, but instead of `Thread.currentThread().getContextClassLoader()`, which probably returns null, you should be passing `this.getClass().getClassLoader()` to the `ServiceLoader.load(...)`.

**Comment by dhpagani** [comment]
That made it !
Thank you @laurit !
